File: 57-insert-interval/insert-interval.js
/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
var insert = function (intervals, newInterval) {
    const res = [];
    let [newStart, newEnd] = newInterval;

    for (const [start, end] of intervals) {
        if (end < newStart) {
            // Current interval ends before the new interval starts
            res.push([start, end]);
        } else if (start > newEnd) {
            // Current interval starts after the new interval ends
            res.push([newStart, newEnd]);
            newStart = start;
            newEnd = end;
        } else {
            // Overlapping intervals; merge them
            newStart = Math.min(newStart, start);
            newEnd = Math.max(newEnd, end);
        }
    }

    // Add the merged interval (if not already added)
    res.push([newStart, newEnd]);

    return res;
};

File: 57-insert-interval/insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> result = new ArrayList<>();
        int newStart = newInterval[0], newEnd = newInterval[1];
        int i = 0, n = intervals.length;

        // Add all intervals that end before the new interval starts
        while (i < n && intervals[i][1] < newStart) {
            result.add(intervals[i]);
            i++;
        }

        // Merge overlapping intervals with the new interval
        while (i < n && intervals[i][0] <= newEnd) {
            newStart = Math.min(newStart, intervals[i][0]);
            newEnd = Math.max(newEnd, intervals[i][1]);
            i++;
        }
        result.add(new int[] { newStart, newEnd });

        // Add the remaining intervals
        while (i < n) {
            result.add(intervals[i]);
            i++;
        }

        // Convert the result list back to a 2D array
        return result.toArray(new int[result.size()][]);
    }
}